src/mlpack/core/tree/binary_space_tree/ub_tree_split_impl.hpp

```diff
@@ -0,0 +1,326 @@
+/**
+ * @file ub_tree_split_impl.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Implementation of UBTreeSplit, a class that splits a node according
+ * to the median address of points contained in the node.
+ */
+#ifndef MLPACK_CORE_TREE_BINARY_SPACE_TREE_UB_TREE_SPLIT_IMPL_HPP
+#define MLPACK_CORE_TREE_BINARY_SPACE_TREE_UB_TREE_SPLIT_IMPL_HPP
+
+#include "ub_tree_split.hpp"
+#include <mlpack/core/tree/bounds.hpp>
+
+namespace mlpack {
+namespace tree {
+
+template<typename BoundType, typename MatType>
+bool UBTreeSplit<BoundType, MatType>::SplitNode(BoundType& bound,
+                                                MatType& data,
+                                                const size_t begin,
+                                                const size_t count,
+                                                size_t& splitCol)
+{
+  constexpr size_t order = sizeof(AddressElemType) * CHAR_BIT;
+  if (begin == 0 && count == data.n_cols)
+  {
+    // Calculate all addresses.
+    InitializeAddresses(data);
+
+    // Probably this is not a good idea. Maybe it is better to get
+    // a number of distinct samples and find the median.
+    std::sort(addresses.begin(), addresses.end(), ComparePair);
+
+    // Rearrange dataset.
+    PerformSplit(data, count);
```

Comment:
I think that the code could be simplified significantly if you instead used `arma::Mat<AddressElemType>` in order to store the addresses, and then used `sort_index()` to get the sorted order, which you could then apply to the data points and the addresses, and use to calculate `oldFromNew` if needed.  I think that would remove the need for `PerformSplit` altogether.